### PR TITLE
Added map_replace_area tool

### PR DESCRIPTION
**Pull request by sctt**
Motivation:
sometimes mappers need to copy a certain area from a map to another (but also to a different position of the same map).
that's very time consuming on complex maps, as tiles and quads layers have to be manually copied one by one.

map_replace_area is designed to automatically replace all the tiles and quads from a specific area to another.
`Usage: map_replace_area <from_map> <from_x> <from_y> <to_map> <to_x> <to_y> <width> <height> <output_map>`

PoW screenshots attached below.

note: for the moment map_replace_area suppose that you are working on the same map  (that's what i needed), that means source and destination maps must have the same layers structure, or else an error is returned.
but i understand it might be useful to also work on totally different maps, in that case users must be able specify layers they want to consider by associating source map layers to destination map layers (by using a config file maybe).
i'm probably going to make another PR to add this enhancement in the future.

PoW:
Screenshot from 2022-03-16 18-08-12
Screenshot from 2022-03-16 18-08-43
Screenshot from 2022-03-16 18-11-50

### CODE REVIEW

File: src/tools/map_replace_area.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	for(int i = 0; i < 2; i++)
+	{
+		if(!InputMaps[i].Open(pStorage, pMapNames[i], IStorage::TYPE_ABSOLUTE))
+		{
+			dbg_msg("map_replace_area", "Error: unable to open map '%s'", pMapNames[i]);
+			return false;
+		}
+	}
+
+        if(!OutputMap.Open(pStorage, pMapNames[2], IStorage::TYPE_ABSOLUTE))
```
Comment (def-) [inline src/tools/map_replace_area.cpp:105]: weird indentation here.

**Review by def- — CHANGES_REQUESTED** [review]
That's cool for cooperatively working on a map. Should we start shipping these kinds of tools? Right now we are only adding map_convert_07, map_diff and map_extract, see CMakeLists.txt:2547

Is this change missing builds? I thought clang-format would catch the minor formatting things.

**Comment by def-** [comment]
Robyt says:

> the file is not included in cmakelist

File: src/tools/map_replace_area.cpp (hunk 1 above)
Comment (sctt) [reply src/tools/map_replace_area.cpp:105]: i don't see any indentation problem, that selection is purposely outside of the previous iteration.
tell me if there's something i'm not seeing, thanks.

**Comment by sctt** [comment]
> Maybe it should ignore groups that don't have parallax 100 in both x and y, since when you copy areas from those groups, the actual position of those areas aren't at the specified coordinates, but instead scaled away/towards the map origin (top-left), depending on the camera position.

good thinking, i didn't consider parallax groups.
it's a pity to just discard those quads as they could be part of the area you are trying to copy, obliging you to manually add them after merging.
maybe it's better to actually simulate a camera on a standard screen, placing it at the center of the area, recalculate parallax quads positions and then check for replacing. that should be enough for small areas.
for bigger ones instead, the camera could actually iterate scans by single portions and perform the same operation, ignoring quads already replaced.

i'll experiment on this and let you know.

**Comment by def-** [comment]
> i'll experiment on this and let you know.

How's it going? Should we submit it as is or wait for this change?

**Comment by sctt** [comment]
> > i'll experiment on this and let you know.
> 
> How's it going? Should we submit it as is or wait for this change?

it is going well, i've managed to handle the parallax problem with success.
i've also took group offsets into consideration for both tile and quad layers.
now i just need to handle clipping: this way layer groups will be completely handled.

expect a commit soon (sadly i can work on this only for few hours a week, but don't worry: i'm active).

**Comment by murpii** [comment]
That's really cool, thanks for working on this!

**Comment by sctt** [comment]
> That's really cool, thanks for working on this!
np dude, that's useful for me too.

sorry guys if i'm taking even more time, but since i'm on the flow i want to improve this as much as possible before focusing on something else.

here's a little feedback:
1. layer groups are now completely handled: that means that offset, clip and parallax are correctly taken into consideration for both quad and tile layers in a very satisfying (and also optimized) way.
2. what i've put into my head right now is to improve flexibility: considering a case in which source and destination map differs in layer group parameters (different offset/clip/parallax) i don't really like to stop the routine and throw an error, that's inconvenient in many situations, i'd rather recalibrate positions (if possible) to adapt the new element to the different configuration (i've already done a couple of experiments on this, seems to work on basic cases, but i need more time to test and finalize this).

anyway, i'm sure that's the last feature i'm going to add before committing and asking for a merge: i'll keep you updated.

**Comment by heinrich5991** [comment]
Status: Waiting for author, if I understand it correctly.

**Comment by sctt** [comment]
ok, tests went all right, the tool seems to behave as expected even in the most bizarre situations.

_just to recap_: this last commit added the capability of correctly handling layer groups parameters (offset, clip, parallax).
even if source and destination maps have different parameters, the tool is able to adapt the situation in order to make the output area look as much as possible as the input one.

i've attached my unit tests so you can have a more precise idea on how the tool is actually expected to behave.
however, i'm also available on discord to answer any conceptual or code-specific question about the algorithm logic (@def- should know my id).

i don't intend to add more features for the moment: **i confirm that from my side we are ready to merge this**.

i just leave a note for the future as i might wish to enhance this even more, like:
- add a mechanism to map source layers with destination layers
- take position envelops into consideration while evaluating quads
- improve quads evaluation by considering them as polygons instead of one-dimensional pivots.
- handle sound layers also

[tests.tar.gz](https://github.com/ddnet/ddnet/files/8939383/tests.tar.gz)

**Comment by sctt** [comment]
@heinrich5991 all done

**Review by def- — APPROVED** [review]

**Comment by def-** [comment]
bors r+
Thanks

**Pull request merged**